### ISSUE TITLE
Satisfaction is being used without "Whistle while you work"

### ISSUE THREAD

**Issue by roswell67** [doxxx/ffxiv-craft-opt-web#163]
Satisfaction needs to consume Whistles, the Whistles are not available without Whistle while you work action.

look at the generated macro below:

Cross-class Macro:
/aaction clear
/aaction "Comfort Zone" on
/aaction "Steady Hand II" on
/aaction "Hasty Touch" on
/aaction "Byregot's Blessing" on
/echo Cross-class action setup complete <se.1>

Macro #1:
/ac "Comfort Zone" <wait.2>
/ac "Inner Quiet" <wait.2>
/ac "Steady Hand II" <wait.2>
/ac "Patient Touch" <wait.3>
/ac "Patient Touch" <wait.3>
/ac "Hasty Touch" <wait.3>
/ac "Hasty Touch" <wait.3>
/ac "Hasty Touch" <wait.3>
/ac "Hasty Touch" <wait.3>
/ac "Satisfaction" <wait.3>
/ac "Master's Mend II" <wait.3>
/ac "Comfort Zone" <wait.2>
/ac "Steady Hand II" <wait.2>
/ac "Hasty Touch" <wait.3>
/echo Macro #1 complete <se.1>

Macro#2:
/ac "Hasty Touch" <wait.3>
/ac "Hasty Touch" <wait.3>
/ac "Hasty Touch" <wait.3>
/ac "Hasty Touch" <wait.3>
/ac "Hasty Touch" <wait.3>
/ac "Great Strides" <wait.2>
/ac "Byregot's Blessing" <wait.3>
/ac "Master's Mend II" <wait.3>
/ac "Careful Synthesis II" <wait.3>
/ac "Hasty Touch" <wait.3>
/ac "Careful Synthesis II" <wait.3>
/ac "Careful Synthesis II" <wait.3>
/ac "Careful Synthesis II" <wait.3>
/ac "Careful Synthesis II" <wait.3>
/ac "Careful Synthesis III" <wait.3>

**Comment by roswell67** [comment]
using the ffxiv-beta version

**Comment by doxxx** [comment]
When filing issues about Solver output, please include the contents of the Execution Log tab on the Solver page, otherwise it can be very hard to pinpoint the problem due to the semi-random nature of the Solver.

But I think I can reproduce the problem in the Simulator manually.

**Comment by doxxx** [comment]
Did you turn off Use Conditions in the Options?

**Comment by doxxx** [comment]
Yeah, turning off Use Conditions will make the solver think it can use Satisfaction with impunity because that bypasses the check for Whistles == 3.

If you want to keep Use Conditions off, I would recommend deselecting the Satisfaction action.

**Comment by doxxx** [comment]
Thank you for the detailed bug report! With Use Conditions on, the bug originally described here won't affect you.

Instead what I believe has happened is that performing two "nothing" actions at that point in the sequence, while Comfort Zone was active, allowed it to regain enough CP to complete the rest of the sequence. Since wasted actions are penalized only slightly when evaluating how good a sequence is, this sequence scored higher than the rest that the Solver came up with.

Something to bear in mind is that the Solver operates randomly. It generates sequences of random actions, evaluates each sequence to see how good it is according to a set of constraints and a model of how actions work, and then mutates the sequences randomly to see if that improves any of them. It doesn't truly understand the process of crafting like you or I would. :)

However, I believe the bug here is that the Solver believes that Satisfaction is usable but does nothing at that point (like Observe), when it sounds like it is actually not usable and doesn't result in the expected CP gains from Comfort Zone.

We don't have a good way to model random effects like Whistle While You Work, Tricks of the Trade, etc. simply because they're random. If the Solver puts them in the sequence, you should consider it as a hint that you're probably going to need to use these actions somewhere in the sequence in order to get the desired result, not necessarily exactly where the Solver placed them.

If you just want to generate macros that are 100% reliable, you should disable the actions that rely on random effects like this.

**Comment by Tubben** [comment]
Thanks for that detailed answer :-). I think most people do use the solver, like i do most of the time if doing some random recipe: Enter item, hit solve, copy makros and build that item without thinking any further. I think the solver should just not use an skill which is independent on some other skill, OR should atleast use that other skill at the start. 

What makes me wonder at that sequence above : Satisfaction regenerate 15 HP and is used twice, but the sequence ended with 31 hp left.
